test/Test/Database/LSMTree/Internal/Integration.hs

```diff
@@ -0,0 +1,199 @@
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal.Integration" [
+      testGroup "InMemLookupData" $ prop_arbitraryAndShrinkPreserveInvariant (deepseqInvariant @InMemLookupData)
+    , localOption (QuickCheckMaxSize 1000) $
+      testProperty "prop_inMemRunLookupAndConstruction" $ prop_inMemRunLookupAndConstruction
+    ]
+
+-- | Construct a run incrementally, then test a number of positive and negative lookups.
+prop_inMemRunLookupAndConstruction :: InMemLookupData -> Property
+prop_inMemRunLookupAndConstruction dat =
+    Map.size runData > 0 ==>
+        tabulateKeySizes
+      $ tabulateValueSizes
+      $ tabulateNumKeyEntryPairs
+      $ tabulateNumPages
+      $ conjoin (fmap check prepped) .&&. conjoin (fmap check' prepped')
+  where
+    InMemLookupData{runData, lookups} = dat
+
+    tabulateKeySizes = tabulate "Size of key in run" [showPowersOf10 $ sizeofKey k | k <- Map.keys runData ]
+    tabulateValueSizes = tabulate "Size of value in run" [showPowersOf10 $ onValue 0 sizeofValue e | e <- Map.elems runData]
+    tabulateNumKeyEntryPairs = tabulate "Number of key-entry pairs" [showPowersOf10 (Map.size runData) ]
+    tabulateNumPages = tabulate "Number of pages" [showPowersOf10 (Map.size ps) | let (ps, _, _) = run]
+
+    run = mkTestRun runData
+    -- prepLookups says that a key /could be/ in the given page
+    prepped = prepLookups [run] lookups
+    -- prepLookups says that a key /is definitely not/ in the given page
+    prepped' = Set.toList (Set.fromList lookups Set.\\ Set.fromList (fmap fst prepped))
+
+    -- Check that a key /is definitely not/ in the given page.
+    check' :: SerialisedKey -> Property
+    check' k =
+          tabulate1Pre (classifyBin (isJust truth) False)
+        $ tabulate1Post (classifyBin (isJust truth) (isJust test))
+        $ Map.lookup k runData === Nothing
+      where truth = Map.lookup k runData
+            test  = Nothing
+
+    -- | Check that a key /could be/ in the given page
+    check :: (SerialisedKey, (Map Int RawPage, PageSpan)) -> Property
+    check (k, (ps, PageSpan (PageNo i) (PageNo j)))
+      | i <= j    = tabulate "PageSpan distribution" [showPowersOf10 $ j - i + 1]
+                  $ tabulate1Pre (classifyBin (isJust truth) True)
+                  $ tabulate1Post (classifyBin (isJust truth) (isJust test))
+                  $ truth === test
+      | otherwise = error "impossible: end of a page span can not come before its start"
+      where
+        (truth, test) | i == j    = ( Map.lookup k runData
+                                    , coerce $ rawPageEntry (ps Map.! i) (coerce k)
+                                    )
+                      | otherwise = ( Map.lookup k runData
+                                    , fmap
+                                        (first concatOverflow)
+                                        (rawPageValue1Prefix (ps Map.! coerce i) (coerce k))
+                                    )
```

Comment:
```suggestion
        truth = Map.lookup k runData
        test | i == j    = coerce $ rawPageEntry (ps Map.! i) (coerce k)
             | otherwise = fmap (first concatOverflow)
                                (rawPageValue1Prefix (ps Map.! coerce i) (coerce k))
```